src/absil/illib.fs

```diff
@@ -448,6 +448,46 @@ module Dictionary =
 module Lazy = 
     let force (x: Lazy<'T>) = x.Force()
 
+//----------------------------------------------------------------------------
+// Singe threaded execution and mutual exclusion
+
+/// Represents a permission active at this point in execution
+type ExecutionToken = interface end
+
+/// Represents a token that indicates execution on the compilation thread, i.e. 
+///   - we have full access to the (partially mutable) TAST and TcImports data structures
+///   - compiler execution may result in type provider invocations when resolving types and members
+///   - we can access various caches in the SourceCodeServices
+///
+/// Like other execution tokens this should be passed via argument passing and not captured/stored beyond
+/// the lifetime of stack-based calls. This is not checked, it is a discipline withinn the compiler code. 
+type CompilationThreadToken() = interface ExecutionToken
+
+/// Represnts a place where we are stating that execution on the compilation thread is required.  The
+/// reason why will be documented in a comment in the code at the callsite.
+let RequireCompilationThread (_ctok: CompilationThreadToken) = ()
+
+/// Represnts a place in the compiler codebase where we are passed a CompilationThreadToken unnecessarily.
+/// This reprents code that may potentially not need to be executed on the compilation thread.
+let DoesNotSpecificallyRequireCompilerThreadAndCouldLikelyBeConcurrent (_ctok: CompilationThreadToken) = ()
```

Comment:
@saul Thanks for reviewing! Any more comments much appreciated

This is  a deliberately verbose and lengthy indicator that further analysis is required.  

Not all uses of mutable state and no type provider invocations have associated RequireCompilationThread calls to force the tokens to be passed along.  So we need to carefully check before taking code off the compilation thread.  

Some good examples are [here](https://github.com/Microsoft/visualfsharp/pull/2371/files#diff-d3c6e872b60cb8a91665cdd9d5e5ca27R1926) and [here](https://github.com/Microsoft/visualfsharp/pull/2371/files#diff-d3c6e872b60cb8a91665cdd9d5e5ca27R1939).  @vasily-kirichenko had previously called out these as unnecessarily using the reactor compilation thread just to collect some information from an otherwise immutable data structure (actually a data structure that was originally mutable but is no longer being written to by this point).  I'm now much more confident that these could be taken off the reactor thread, but we need to double-check in a separate PR.

Perhaps the name ``DoesNotRequireCompilerThreadTokenAndCouldPossiblyBeMadeConcurrent`` - I don't mind the length as long as it's clear.